internal/wire/ack_frame.go

```diff
@@ -9,340 +9,166 @@ import (
-				frame.AckRanges[len(frame.AckRanges)-1].Last -= protocol.PacketNumber(gap)
-			} else {
-				lastRangeComplete = false
-				ackRange := AckRange{
-					Last: frame.AckRanges[len(frame.AckRanges)-1].First - protocol.PacketNumber(gap) - 1,
-				}
-				ackRange.First = ackRange.Last - length + 1
-				frame.AckRanges = append(frame.AckRanges, ackRange)
-			}
-
-			if length > 0 {
-				lastRangeComplete = true
-			}
-
-			inLongBlock = (ackBlockLength == 0)
+		gap := protocol.PacketNumber(g)
+		if smallest < gap+2 {
+			return nil, errInvalidAckRanges
 		}
+		largest := smallest - gap - 2
 
-		// if the last range was not complete, First and Last make no sense
-		// remove the range from frame.AckRanges
-		if !lastRangeComplete {
-			frame.AckRanges = frame.AckRanges[:len(frame.AckRanges)-1]
+		ab, err := utils.ReadVarInt(r)
+		if err != nil {
+			return nil, err
 		}
+		ackBlock := protocol.PacketNumber(ab)
 
-		frame.LowestAcked = frame.AckRanges[len(frame.AckRanges)-1].First
-	} else {
-		if frame.LargestAcked == 0 {
-			frame.LowestAcked = 0
-		} else {
-			frame.LowestAcked = protocol.PacketNumber(largestAcked + 1 - ackBlockLength)
+		if ackBlock > largest {
+			return nil, errInvalidAckRanges
 		}
+		smallest = largest - protocol.PacketNumber(ackBlock)
+		frame.AckRanges = append(frame.AckRanges, AckRange{First: smallest, Last: largest})
 	}
 
+	frame.LowestAcked = smallest
 	if !frame.validateAckRanges() {
-		return nil, ErrInvalidAckRanges
+		return nil, errInvalidAckRanges
 	}
 
-	var numTimestamp byte
-	numTimestamp, err = r.ReadByte()
-	if err != nil {
-		return nil, err
-	}
-
-	if numTimestamp > 0 {
-		// Delta Largest acked
-		_, err = r.ReadByte()
-		if err != nil {
-			return nil, err
-		}
-		// First Timestamp
-		_, err = utils.GetByteOrder(version).ReadUint32(r)
-		if err != nil {
-			return nil, err
-		}
-
-		for i := 0; i < int(numTimestamp)-1; i++ {
-			// Delta Largest acked
-			_, err = r.ReadByte()
-			if err != nil {
-				return nil, err
-			}
-
-			// Time Since Previous Timestamp
-			_, err = utils.GetByteOrder(version).ReadUint16(r)
-			if err != nil {
-				return nil, err
-			}
-		}
-	}
 	return frame, nil
 }
 
 // Write writes an ACK frame.
 func (f *AckFrame) Write(b *bytes.Buffer, version protocol.VersionNumber) error {
-	largestAckedLen := protocol.GetPacketNumberLength(f.LargestAcked)
-
-	typeByte := uint8(0x40)
-
-	if largestAckedLen != protocol.PacketNumberLen1 {
-		typeByte ^= (uint8(largestAckedLen / 2)) << 2
+	if !version.UsesIETFFrameFormat() {
+		return f.writeLegacy(b, version)
 	}
 
-	missingSequenceNumberDeltaLen := f.getMissingSequenceNumberDeltaLen()
-	if missingSequenceNumberDeltaLen != protocol.PacketNumberLen1 {
-		typeByte ^= (uint8(missingSequenceNumberDeltaLen / 2))
-	}
+	b.WriteByte(0xe)
+	utils.WriteVarInt(b, uint64(f.LargestAcked))
+	utils.WriteVarInt(b, encodeAckDelay(f.DelayTime))
 
+	// TODO: limit the number of ACK ranges, such that the frame doesn't grow larger than an upper bound
```

Comment:
I'm still fine with introducing an arbitrary cutoff. Maybe we should raise this in the IETF though.

Reply:
What exactly do you want to raise there?

Reply:
Whether the spec should define a max number of ACK ranges.